trollmoves/client.py

```diff
@@ -312,57 +315,116 @@ def replace_mda(msg, kwargs):
+    return requester.send_and_recv(req, timeout=timeout), hostname
 
-    if response and response.type in ['file', 'collection', 'dataset']:
-        LOGGER.debug("Server done sending file")
-        with cache_lock:
-            for uid in gen_dict_extract(msg.data, 'uid'):
-                file_cache.append(uid)
-        try:
-            lmsg = unpack_and_create_local_message(response, local_dir, unpack, delete)
-        except IOError:
-            LOGGER.exception("Couldn't unpack %s", str(response))
-            return
-        if publisher:
-            lmsg = make_uris(lmsg, destination, login)
-            lmsg.data['origin'] = response.data['request_address']
-            lmsg.data.pop('request_address', None)
-            lmsg = replace_mda(lmsg, kwargs)
-            lmsg.data.pop('destination', None)
 
-            LOGGER.debug("publishing %s", str(lmsg))
-            publisher.send(str(lmsg))
+def send_ack(msg, timeout):
+    """Send an ACK (no push required)."""
+    req = Message(msg.subject, 'ack', data=msg.data)
+    LOGGER.debug("Sending: %s", str(req))
 
-    elif response and response.type == "ack":
+    response, hostname = send_request(msg, req, timeout)
+
+    if response and response.type == "ack":
         pass
     else:
         LOGGER.error("Failed to get valid response from server %s: %s",
                      str(hostname), str(response))
 
 
-def reload_config(filename, chains, callback=request_push, pub_instance=None):
-    """Rebuild chains if needed (if the configuration changed) from *filename*.
-    """
+def terminate_transfers(uid, timeout):
+    """Send ACK to remaining sources for uid and remove from the ongoing tranfers list."""
+    with ongoing_transfers_lock:
+        msgs = ongoing_transfers.pop(uid, [])
+    for msg in msgs:
+        send_ack(msg, timeout)
+
+
+def get_msg_uid(msg):
+    """Compute the uid of the message."""
+    filenames = sorted(gen_dict_extract(msg.data, 'uid'))
+    m = hashlib.md5()
+    for filename in filenames:
+        m.update(filename.encode('utf-8'))
+    return m.hexdigest()
+
 
+def iterate_messages(uid):
+    """Iterate over all messages for a uid."""
+    while True:
+        try:
+            msg = get_next_msg(uid)
+        except IndexError:
+            raise StopIteration
+        yield msg
+
+
+def get_next_msg(uid):
+    """Get the next message with this *uid* from the available sources."""
+    with ongoing_transfers_lock:
+        return ongoing_transfers[uid].pop(0)
+
+
+def request_push(msg, destination, login, publisher=None, unpack=None, delete=False, **kwargs):
+    """Request a push for data."""
+    huid = get_msg_uid(msg)
+    with ongoing_transfers_lock:
+        if huid in ongoing_transfers:
+            ongoing_transfers[huid].append(msg)
+            return
+        else:
+            ongoing_transfers[huid] = [msg]
+
+    if already_received(msg):
+        timeout = float(kwargs["req_timeout"])
+        resend_if_local(msg, publisher)
```

Comment:
@pnuu @TAlonglong would you be ok if I removed this line ? This basically resends the message if the file is in the file_cache (meaning it has already been processed) and that the files are pointing to the local machine. I don't think this is actually needed. What do you think ?

Reply:
I'm ok with that. But note that I have only ran few tests so far, so don't put too much weight on my thougts :grin: Anyway, sounds like removing it won't be affecting our production in the future either.

Reply:
Sorry, I dont think I get the logic. Why would you resend the message if its already received and is local? Its a lots of thinks I dont get in trollmove, so maybe I just stay away :-)

Reply:
No, you are right. I think it's some legacy stuff that I forgot to remove. I'll clean this up now.